### ISSUE TITLE
False positive for clone_on_copy

### ISSUE THREAD

**Issue by vallentin** [rust-lang/rust-clippy#4384]
I've encountered a false positive, when calling `.clone()` on a field of a struct. Followed by calling a method (`&self`) that takes a closure, which borrows the original struct.

*I encountered it using the [`glfw`](https://crates.io/crates/glfw) crate, and boils down to the following:*

```rust
// glfw: Glfw + Copy
window.glfw.clone().with_primary_monitor(|_, _| {
    println!("{:?}", window.get_size());
});
```

Clippy suggests removing `.clone()` since `Glfw` implements `Copy`.

> using `clone` on a `Copy` type
>
> help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#clone_on_copy
> help: try removing the `clone` call: `window.glfw`clippy(clippy::clone_on_copy)

However, that doesn't work, since it would *reborrow* `window`.

> error[E0502]: cannot borrow `window.glfw` as mutable because it is also borrowed as immutable

---

Here's two minimal, reproducible examples.

#### Example 1

```rust
let mut i = 0;

fn foo<F>(_: &i32, mut f: F) where F: FnMut() {
    f();
}

foo(&i.clone(), || {
    i += 1;
});

println!("{}", i);
```

#### Example 2

```rust
#[derive(Clone, Copy, Debug)]
struct Foo(i32);

impl Foo {
    fn bar<F>(&self, mut f: F) where F: FnMut() {
        f();
    }
}

let mut foo = Foo(0);

foo.clone().bar(|| {
    foo.0 += 1;
});

println!("{:?}", foo);
```

*I'm aware that changing `&self` to `self` and removing `.clone()` works. However, that would require me to modify `glfw`.*

Version: clippy 0.0.212 (082cfa79 2019-06-07)
